azure-common/azure/common/client_factory.py

```diff
@@ -49,72 +60,125 @@ def get_client_from_cli_profile(clientclass, **kwargs):
+        authority_url = (config_dict['activeDirectoryEndpointUrl'] + '/' + 
+                         config_dict['tenantId'])
         context = adal.AuthenticationContext(authority_url, api_version=None)
         parameters['credentials'] = AdalAuthentication(
             context.acquire_token_with_client_credentials,
-            config_dict['managementURI'],
-            config_dict['client'],
-            config_dict['key']
+            config_dict['managementEndpointUrl'],
+            config_dict['clientId'],
+            config_dict['clientSecret']
         )
 
     parameters.update(kwargs)
-    return clientclass(**parameters)
+    return _instanciate_client(clientclass, **parameters)
+
+def get_client_from_auth_file(clientclass, auth_path=None, **kwargs):
+    """Return a SDK client initialized with auth file.
+
+    The easiest way to obtain this file is to call the following CLI commands:
+
+    .. code:: bash
+
+        az ad sp create-for-rbac --sdk-auth
+
+    You can specific the file path directly, or fill the environment variable AZURE_AUTH_LOCATION.
+    File must be UTF-8.
+
+    This method will fill automatically the following client parameters:
+    - credentials
+    - subscription_id
+    - base_url
+
+    Parameters provided in kwargs will override parameters and be passed directly to the client.
+
+    :Example:
+
+    .. code:: python
+
+        from azure.common.client_factory import get_client_from_auth_file
+        from azure.mgmt.compute import ComputeManagementClient
+        client = get_client_from_auth_file(ComputeManagementClient)
+
+    Example of file:
+
+    .. code:: json
+
+        {
+            "clientId": "ad735158-65ca-11e7-ba4d-ecb1d756380e",
+            "clientSecret": "b70bb224-65ca-11e7-810c-ecb1d756380e",
+            "subscriptionId": "bfc42d3a-65ca-11e7-95cf-ecb1d756380e",
+            "tenantId": "c81da1d8-65ca-11e7-b1d1-ecb1d756380e",
+            "activeDirectoryEndpointUrl": "https://login.microsoftonline.com",
+            "resourceManagerEndpointUrl": "https://management.azure.com/",
+            "activeDirectoryGraphResourceId": "https://graph.windows.net/",
+            "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/",
+            "galleryEndpointUrl": "https://gallery.azure.com/",
+            "managementEndpointUrl": "https://management.core.windows.net/"
+        }
+
+    .. versionadded:: 1.1.7
+
+    :param clientclass: A SDK client class
+    :param str auth_path: Path to the file.
+    :return: An instanciated client
+    :raises: KeyError if AZURE_AUTH_LOCATION is not an environment variable and no path is provided
+    :raises: FileNotFoundError if provided file path does not exists
+    :raises: json.JSONDecodeError if provided file is not JSON valid
+    :raises: UnicodeDecodeError if file is not UTF8 compliant
+    """
+    auth_path = auth_path or os.environ['AZURE_AUTH_LOCATION']
+
+    with io.open(auth_path, 'r', encoding='utf-8') as auth_fd:
```

Comment:
So several things:
- JSON is supposed to be UTF-8
- The actual file encoding is decided by the customer, and the default encoding of his shell, since it will probably create the file using a stdout redirection (`az ... --sdk-auth > mycredentials.json`)
- I did some testing, and `utf-8-sig` seems to be able to decode `utf8` string, but not the opposite:

```shell
>>> "tata".encode("utf-8-sig").decode("utf-8")
u'\ufefftata'
>>> "tata".encode("utf-8").decode("utf-8-sig")
u'tata'
```

So I think the best approach is to use "utf-8-sig", which will be most compatible and support ASCII, utf8 and UTF8 with BOM at the same time. But I won't move from utf8 and try to detect cpXXXX with chardet.